keps/sig-cli/2906-kustomize-function-catalog/README.md

```diff
@@ -0,0 +1,667 @@
+
+In order to use Kustomize plugins today, an end user must explicitly provide a reference to the plugin implementation. For example, consider the use of a plugin with a Kustomization. First, the user would define a resource configuration as follows:
+
+```yaml
+apiVersion: team.example.com/v1alpha1
+kind: HTTPLoadBalancer
+metadata:
+  name: lb
+  annotations:
+    config.kubernetes.io/function: |
+      container:
+        image: docker.example.com/kustomize-modules/lb:v0.1.1
+spec:
+  selector:
+    matchLabels:
+      app: nginx-example
+  expose:
+    serviceName: nginx
+    port: 80
+```
+
+This is then referenced from a Kustomization in the following way:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+resources:
+- lb.yaml
+```
+
+This explicit configuration requires the user to include two separate pieces of configuration to use the function: 
+* they must provide the API Version and Kind
+* they must provide an explicit reference to the Docker container that will be used 
+
+When using a container based plugin provider with the `Composition` API, the user must still specify this information:
+
+```yaml
+# app/composition.yaml
+kind: Composition
+
+modules:
+# generate resources for a Java application
+- apiVersion: example.com/v1
+  kind: JavaApplication
+  provider: {container: {image: example/module_providers/java:v1.0.0}}
+  spec:
+    application: team/my-app
+    version: v1.0
+```
+
+Once the explicit container reference is provided, Kustomize is able to download and run this image as part of the Kustomize build step, by invoking the user installed Docker client and leveraging local images or OCI registries. In addition to container based plugins, ad-hoc functions can also currently be written using the [Starlark programming language](https://github.com/bazelbuild/starlark) and other non-container based mechanisms. Unlike container based plugins, these plugins do not currently have an associated registry concept and they must be stored locally. Discovery and installation of these providers are both currently left to the users. 
+
+In addition to these user defined plugins, this KEP is also partially motivated by a need to change how Kustomize provides officially supported functionality. Currently, to support a given piece of functionality officially a built-in function must be created (and typically also added to the Kustomization API). Some of these features would be better implemented as extensions for security reasons or to limit the dependency graph of Kustomize and the integration with kubectl. No mechanism currently exists, however, to support official distribution of these capabilities, so they are instead built into Kustomize.
```

Comment:
one of the other reasons is that one might want to mix and match the kustomize binary version with the plugin version.  This frees up the core kustomize team to introduce breaking changes that can clean up legacy design decisions without having to rev absolutely everything.